src/repository.py

```diff
@@ -0,0 +1,221 @@
+
+import base64
+import logging
+import re
+from pathlib import Path
+
+from git import GitCommandError
+from git.repo import Repo
+from github import Github
+from github.GithubException import GithubException
+from github.Repository import Repository
+
+from .constants import DOCUMENTATION_FOLDER_NAME
+from .exceptions import InputError, RepositoryClientError
+
+GITHUB_HOSTNAME = "github.com"
+HTTPS_URL_PATTERN = re.compile(rf"^https?:\/\/.*@?{GITHUB_HOSTNAME}\/(.+\/.+?)(.git)?$")
+ACTIONS_USER_NAME = "upload-charms-docs-bot"
+ACTIONS_USER_EMAIL = "upload-charms-docs-bot@users.noreply.github.com"
+ACTIONS_PULL_REQUEST_TITLE = "[upload-charm-docs] Migrate charm docs"
+ACTIONS_PULL_REQUEST_BODY = (
+    "This pull request was autogenerated by upload-charm-docs to migrate "
+    "existing documentation from server to the git repository."
+)
+PR_LINK_NO_CHANGE = "<not created due to no changes in repository>"
+
+CONFIG_USER_SECTION_NAME = "user"
+CONFIG_USER_NAME = (CONFIG_USER_SECTION_NAME, "name")
+CONFIG_USER_EMAIL = (CONFIG_USER_SECTION_NAME, "email")
+
+
+class Client:
+    """Wrapper for git/git-server related functionalities."""
+
+    def __init__(self, repository: Repo, github_repository: Repository) -> None:
+        """Construct.
+
+        Args:
+            repository: Client for interacting with local git repository.
+            github_repository: Client for interacting with remote github repository.
+        """
+        self._git_repo = repository
+        self._github_repo = github_repository
+        self._configure_git_user()
+
+    def _configure_git_user(self) -> None:
+        """Configure action git profile defaults.
+
+        Configured profile appears as the git committer.
+        """
+        config_reader = self._git_repo.config_reader(config_level="repository")
+        with self._git_repo.config_writer(config_level="repository") as config_writer:
+            if not config_reader.has_section(
+                CONFIG_USER_SECTION_NAME
+            ) or not config_reader.get_value(*CONFIG_USER_NAME):
+                config_writer.set_value(*CONFIG_USER_NAME, ACTIONS_USER_NAME)
+            if not config_reader.has_section(
+                CONFIG_USER_SECTION_NAME
+            ) or not config_reader.get_value(*CONFIG_USER_EMAIL):
+                config_writer.set_value(*CONFIG_USER_EMAIL, ACTIONS_USER_EMAIL)
+
+    def check_branch_exists(self, branch_name: str) -> bool:
+        """Check if branch exists on remote.
+
+        Args:
+            branch_name: Branch name to check on remote.
+
+        Raises:
+            RepositoryClientError: if unexpected error occurred during git operation.
+
+        Returns:
+            True if branch already exists, False otherwise.
+        """
+        try:
+            self._git_repo.git.fetch("origin", branch_name)
+            return True
+        except GitCommandError as exc:
+            if "couldn't find remote ref" in exc.stderr:
```

Comment:
This forces user to name the remote `origin` couldn't this cause issues ?
That's probably a edge case... what do you think ?

Reply:
I think the way the git repo is initialised is entirely within the control of the action, `origin` is just the default name of the upstream set in git. Any examples you can think of where using `origin` could be problematic?

Reply:
This is also covered by the "this code was just copied", although if there is a problem it is worth discussing and fixing any bugs ASAP

Reply:
Well indeed, the repo is initialized by the action so that won't cause any problem as long as the default will be "origin", perhaps making it a constant as it's repeated multiple times ?
Or if that's not mandatory to specify the remote (by default it should use the default remote) then just remove "origin" ?